### PR TITLE
reorganize repo structure

### PR DESCRIPTION
**Pull request by chenyan-dfinity**
Fix https://github.com/dfinity/candid/issues/59

* Update README
* Put everything under three top-level directories: `spec/`, `rust/`, and `docs/`

### CODE REVIEW

File: spec/Candid-Motoko.md [hunk 1]
```diff
@@ -1,32 +1,32 @@
-The IDL-Motoko integration
```
Comment (rossberg) [inline spec/Candid-Motoko.md:1]: Sorry, I hadn't seen this before. I actually think that this file should stay in the Motoko repo. It's not part of the Candid spec.

File: README.md [hunk 2]
```diff
@@ -1,18 +1,36 @@
...
-# Contribution
+## Implementations
+
+Candid supports several different programming languages. 
+This repository contains some of the implementations developed by DFINITY.
+
+* [Rust](rust/): A serialization library based on Serde, and a compiler for generating bindings for other languages.
+* Motoko: Compiler support for importing/export Candid files.
+* JavaScript: As part of the [JavaScript User Library](https://www.npmjs.com/package/@dfinity/agent), we provide a library for serialization of native JavaScript values, and a visitor class for extending Candid for building generic tools such as UI and random testing.
+* More to come...
```
Comment (rossberg) [inline README.md:21]: I'd remove that, as I don't think we have any specific plans at this point.

File: README.md [hunk 3]
```diff
@@ -1,18 +1,36 @@
...
+
+A list of community maintained Candid libraries:
+
+* [Haskell](https://github.com/nomeata/haskell-candid)
+* [Elm](https://github.com/chenyan2002/ic-elm/)
+* [Kotlin](https://github.com/seniorjoinu/candid-kt)
```
Comment (nomeata) [inline README.md:27]: Oh wow, I didn't konw about this. This makes providing candid test data even more relevant (even a simple one).

File: spec/Candid-Motoko.md (hunk 1 above)
Comment (chenyan-dfinity) [reply spec/Candid-Motoko.md:1]: It's arguable, and can stay in either repo. I think it has value to stay here, as it provides a good example of how Candid integrates with other languages, and how to deal with type gaps. So I slightly prefer it to stay here. WDYT?

Comment (nomeata) [reply spec/Candid-Motoko.md:1]: Maybe we should have one for Rust and JS (where the implemenation is in this repo), as good examples :-)

Comment (chenyan-dfinity) [reply spec/Candid-Motoko.md:1]: The Rust one is harder to write, as it needs to fight between nominal typing and structural typing. Motoko-Candid is actually the simplistic example.

**Comment by matthewhammer** [comment]
I'm happy to have a `tools` directory so long as it doesn't imply that each Rust library crate with a binary tool (e.g., `candiff`) must be artificially separated into two distinct Rust crates.  That's annoying, if only because we need two distinct names in the Rust package ecosystem.

However, I see the value in a toplevel `tools` directory, that gives a comprehensive view of all CLI tools for Candid, regardless of how they are implemented.

Perhaps I can put shell scripts there that point at the (eventual) binary targets of the Rust crates?

**Comment by chenyan-dfinity** [comment]
It's unclear to me what should be put under `tools/`. We can also put the whole `candiff` code under `tools/`, but tools are not necessarily written in Rust. Should we have `tools/rust/`, `tools/javascript/`, or `rust/tools`, `javascript/tools`?

One thing I would like to be under `tools/` is some kind of test suites/framework that can test different implementations. `candiff` seems to be able to help with the test suites, but maybe we only need the binary, not the library code.

**Comment by matthewhammer** [comment]
> It's unclear to me what should be put under tools/

Also not clear to me, for the reasons you said.  One possible answer is to have it just be symbolic links, or one-line shell scripts, or just a README with directions to the other directories, etc.

**Comment by rossberg** [comment]
> It's unclear to me what should be put under tools/

It's very simple: any tool whose _use_ is not specific to Rust. I think you guys are heavily conflating implementation language with target language. The toplevel structure of this repo should reflect the latter, not the former.

For example, a binding generator for Rust clearly is Rust-specific, so should go into rust/ -- regardless of what language it is implemented in! It could be written in OCaml.

A tool like candiff is not specific to any language, so should go under tools. Again regardless of what language it is written in.

Libraries to be used in a specific language are actually a third kind of area. They could go under rust/lib or something.

File: spec/Candid-Motoko.md (hunk 1 above)
Comment (chenyan-dfinity) [reply spec/Candid-Motoko.md:1]: @rossberg Do you want to remove this from the repo? I'm fine either way, but need a decision to merge this PR.

Comment (rossberg) [reply spec/Candid-Motoko.md:1]: [Sorry, day off last Friday.] Yes, I think this should be removed here, but that can be a separate PR.

**Comment by matthewhammer** [comment]
>  a binding generator for Rust clearly is Rust-specific, so should go into rust/ -- regardless of what language it is implemented in! It could be written in OCaml.

Oh, thanks for clarifying.  Indeed, I was confused in the way you said: I was thinking that the language directory indicated the implementation language, rather than the "input language" (nit: I wouldn't call it the "target language", since these tools may not have a meaningful notion of that term.  They all should accept input, if even just to check it somehow.)

**Pull request merged**